keps/sig-cloud-provider/0029-vendor-cinder-driver.md

```diff
@@ -0,0 +1,136 @@
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories [optional]](#user-stories-optional)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+As part of the multi-stage process to remove the OpenStack provider from upstream
+Kubernetes, existing dependencies on the in-tree Cinder driver need to be vendored
+to give parity between the external and in-tree OpenStack provider code.
+
+This is an intermediary step for full removal of the code, and can serve as an
+example for how other providers can approach similar problems in provider code that
+is deeply entangled in upstream code.
+
+## Motivation
+
+As part of the SIG-Cloud-Provider charter to foster neutral collaboration within
+the Kubernetes community, there is an active goal of removing all in-tree cloud
+providers from within the Kuberntes code-base and replacing them with external,
+expert maintained provider code. Over the course of Kubernetes development, the
+in-tree providers have built up a complex set of dependencies that makes full
+replacement challenging. In the case of OpenStack, one of these dependencies is
+with in-tree Cinder integrations.
+
+### Goals
+
+As part of the process to extract and remove the in-tree provider, we need to
+create a path for continuing to support in-tree Cinder dependencies. The
+primary goal of this KEP is to create an interface to the in-tree Cinder provider
+that vendors out the the external cinder provider. This will assist in the larger
+goal of deleting in-tree OpenStack code and allowing projects that depend on the
+provider to continue to work.
+
+A secondary goal is to create a development entrypoint to continue to remove
+in-tree provider code, with a future focus on facilitating the extraction of
+integrated Cinder code.
+
+A tertiary goal is to demonstrate a means for staging removal of in-tree code
+by creating interfaces that external provider can vendor against, with the goal
+of eventual removal of all provider code.
+
+This KEP will be considered completed if the dependencies on the in-tree Cinder
+interface continues to work while calline out to the external provider.
+
+
+### Non-Goals
+
+* Removal of in-tree Cinder code is out of scope for this KEP.
+* Use of the CSI interface is out of scope for thie KEP.
+
+## Proposal
+
+An in-tree interface will be created for the Kubernetes Cinder code. This
+interface will call out to the external Cinder provider, allowing for the
```

Comment:
@hogepodge it's not the `external` provider, we are just moving the code to another repo and just vendoring that repo into main k/k.

the issue here is there is a recursion k/k repends on k/c-p-o and k/c-p-o depends on k/k.... this situation will get better over time as @andrewsykim and others are extracting the code into staging/ repos (into separate repos). eventually this recursion will be no more. but until then we have to be careful when changes are being made in both repos